File: extensions/template-plugin-python/tests/test_pack_tarball.py
"""Phase 31.4 — end-to-end test of `scripts/pack-tarball-python.sh`.

Asserts the bash pipeline produces a tarball whose name + layout
+ sha256 sidecar exactly match the convention 31.1 consumes.

Synthetic SDK: a tempdir with a single `__init__.py` (one line)
substitutes for the real SDK so the test does not depend on the
in-tree SDK source layout. `requirements.txt` is empty, so no
real `pip install` runs (`SKIP_PIP=1` env override).
"""

import hashlib
import os
import shutil
import subprocess
import tarfile
import tempfile
import unittest
from pathlib import Path

TEMPLATE_ROOT = Path(__file__).resolve().parent.parent
PLUGIN_ID = "template_plugin_python"
PLUGIN_VERSION = "0.1.0"


class PackTarballTests(unittest.TestCase):
    def test_pack_tarball_produces_canonical_layout(self):
        with tempfile.TemporaryDirectory() as work_str, tempfile.TemporaryDirectory() as sdk_str, tempfile.TemporaryDirectory() as extract_str:
            work = Path(work_str)
            sdk = Path(sdk_str)
            extract_dir = Path(extract_str)

            # 1. Synthetic SDK with a single import-able file.
            sdk_pkg = sdk / "nexo_plugin_sdk"
            sdk_pkg.mkdir()
            (sdk_pkg / "__init__.py").write_text("# stub SDK\n")

            # 2. Copy template fixture (manifest + scripts + src)
            #    into a fresh work dir so the test does not write
            #    into the workspace.
            for name in ("nexo-plugin.toml", "requirements.txt"):
                shutil.copy(TEMPLATE_ROOT / name, work / name)
            shutil.copytree(TEMPLATE_ROOT / "src", work / "src")
            shutil.copytree(TEMPLATE_ROOT / "scripts", work / "scripts")

            # 3. Run the pack script with SDK_SRC + SKIP_PIP overrides.
            env = dict(os.environ)
            env["SDK_SRC"] = str(sdk_pkg)
            env["SKIP_PIP"] = "1"
            result = subprocess.run(
                ["bash", "scripts/pack-tarball-python.sh"],
                cwd=work,
                env=env,
                capture_output=True,
                check=False,
            )
            self.assertEqual(
                result.returncode,
                0,
                msg=f"pack failed:\nstdout={result.stdout!r}\nstderr={result.stderr!r}",
            )

            # 4. Asset present.
            asset = work / "dist" / f"{PLUGIN_ID}-{PLUGIN_VERSION}-noarch.tar.gz"
            sidecar = work / "dist" / f"{PLUGIN_ID}-{PLUGIN_VERSION}-noarch.tar.gz.sha256"
            self.assertTrue(asset.is_file(), f"asset missing: {asset}")
            self.assertTrue(sidecar.is_file(), f"sha sidecar missing: {sidecar}")

            sidecar_hex = sidecar.read_text().strip()
            self.assertEqual(len(sidecar_hex), 64, f"sha sidecar must be 64 hex chars, got {sidecar_hex!r}")
            self.assertTrue(
                all(c.isdigit() or c in "abcdef" for c in sidecar_hex),
                "sidecar must be lowercase hex",
            )

            # 5. Recompute sha256.
            sha = hashlib.sha256(asset.read_bytes()).hexdigest()
            self.assertEqual(sha, sidecar_hex, "sha256 mismatch vs sidecar")

            # 6. Re-extract and verify layout.
            with tarfile.open(asset, "r:gz") as tf:
                tf.extractall(extract_dir)

            expected_root = {"bin", "lib", "nexo-plugin.toml"}
            actual_root = {p.name for p in extract_dir.iterdir()}
            self.assertEqual(
                actual_root,
                expected_root,
                f"unexpected top-level entries: {actual_root - expected_root}",
            )
            self.assertTrue((extract_dir / "bin" / PLUGIN_ID).is_file())
            self.assertTrue((extract_dir / "lib" / "plugin" / "main.py").is_file())
            self.assertTrue(
                (extract_dir / "lib" / "nexo_plugin_sdk" / "__init__.py").is_file()
            )

            # Launcher is executable on Unix.
            mode = (extract_dir / "bin" / PLUGIN_ID).stat().st_mode & 0o777
            self.assertEqual(mode, 0o755, f"launcher mode should be 0o755, got {mode:o}")


if __name__ == "__main__":
    unittest.main()

File: extensions/template-plugin-php/tests/test_pack_tarball.php
<?php

declare(strict_types=1);

/**
 * Phase 31.5.c — end-to-end test of `scripts/pack-tarball-php.sh`.
 *
 * Asserts the bash pipeline produces a tarball whose name +
 * layout + sha256 sidecar match the convention 31.1 consumes.
 *
 * Synthetic vendor: a tempdir with a minimal `vendor/autoload.php`
 * + a stub `vendor/nexo/plugin-sdk/src/PluginAdapter.php`
 * substitutes for a real `composer install` run. SKIP_COMPOSER=1
 * env override bypasses the Composer step.
 */

const PLUGIN_ID = 'template_plugin_php';
const PLUGIN_VERSION = '0.1.0';

function fail(string $msg): never
{
    fwrite(STDERR, "FAIL: $msg\n");
    exit(1);
}

$templateRoot = dirname(__DIR__);

function rcopy(string $src, string $dst): void
{
    if (is_file($src)) {
        @mkdir(dirname($dst), 0755, true);
        if (!copy($src, $dst)) {
            fail("copy $src → $dst failed");
        }
        return;
    }
    if (!is_dir($src)) {
        return;
    }
    @mkdir($dst, 0755, true);
    foreach (scandir($src) as $entry) {
        if ($entry === '.' || $entry === '..') {
            continue;
        }
        rcopy("$src/$entry", "$dst/$entry");
    }
}

$work = sys_get_temp_dir() . '/pack-php-' . bin2hex(random_bytes(6));
$sdk  = sys_get_temp_dir() . '/sdk-stub-' . bin2hex(random_bytes(6));
$extr = sys_get_temp_dir() . '/pack-php-extract-' . bin2hex(random_bytes(6));
mkdir($work, 0755, true);
mkdir($sdk, 0755, true);
mkdir($extr, 0755, true);

// 1. Synthetic SDK stub.
mkdir("$sdk/src", 0755, true);
file_put_contents("$sdk/src/PluginAdapter.php", "<?php // stub SDK\n");
file_put_contents(
    "$sdk/composer.json",
    json_encode([
        'name' => 'nexo/plugin-sdk',
        'type' => 'library',
        'autoload' => ['psr-4' => ['Nexo\\Plugin\\Sdk\\' => 'src/']],
    ], JSON_PRETTY_PRINT),
);

// 2. Copy template fixture into work dir (manifest + scripts + src).
copy("$templateRoot/nexo-plugin.toml", "$work/nexo-plugin.toml");
rcopy("$templateRoot/src", "$work/src");
rcopy("$templateRoot/scripts", "$work/scripts");

// 3. Provide a synthetic vendor/ tree so SKIP_COMPOSER=1 has
//    something to pack. Mirrors what `composer install --no-dev
//    --optimize-autoloader --classmap-authoritative` would
//    produce.
mkdir("$work/vendor/composer", 0755, true);
file_put_contents("$work/vendor/autoload.php", "<?php // stub autoloader\n");
file_put_contents("$work/vendor/composer/autoload_classmap.php", "<?php return [];\n");
mkdir("$work/vendor/nexo/plugin-sdk/src", 0755, true);
file_put_contents("$work/vendor/nexo/plugin-sdk/src/PluginAdapter.php", "<?php // vendored SDK\n");

// 4. Run pack with SDK_SRC + SKIP_COMPOSER overrides.
$env = array_merge($_ENV, [
    'SDK_SRC' => $sdk,
    'SKIP_COMPOSER' => '1',
    'PATH' => getenv('PATH'),
]);
$cwd = $work;
$proc = proc_open(
    ['bash', 'scripts/pack-tarball-php.sh'],
    [0 => ['pipe', 'r'], 1 => ['pipe', 'w'], 2 => ['pipe', 'w']],
    $pipes,
    $cwd,
    $env,
);
if (!is_resource($proc)) {
    fail('proc_open pack-tarball failed');
}
fclose($pipes[0]);
$stdout = stream_get_contents($pipes[1]);
$stderr = stream_get_contents($pipes[2]);
fclose($pipes[1]);
fclose($pipes[2]);
$code = proc_close($proc);
if ($code !== 0) {
    fail("pack failed: code=$code stdout=$stdout stderr=$stderr");
}

// 5. Asset present.
$assetName = sprintf('%s-%s-noarch.tar.gz', PLUGIN_ID, PLUGIN_VERSION);
$asset = "$work/dist/$assetName";
$sidecar = "$work/dist/$assetName.sha256";
if (!is_file($asset)) {
    fail("asset missing: $asset");
}
if (!is_file($sidecar)) {
    fail("sha sidecar missing: $sidecar");
}

// 6. Sidecar is 64 lowercase hex chars.
$sidecarHex = trim(file_get_contents($sidecar));
if (strlen($sidecarHex) !== 64 || preg_match('/^[0-9a-f]{64}$/', $sidecarHex) !== 1) {
    fail("sidecar must be 64 lowercase hex chars, got: $sidecarHex");
}

// 7. Recompute sha256.
$computed = hash_file('sha256', $asset);
if ($computed !== $sidecarHex) {
    fail("sha256 mismatch: computed=$computed sidecar=$sidecarHex");
}

// 8. Re-extract via system tar + verify layout.
exec("tar -xzf " . escapeshellarg($asset) . " -C " . escapeshellarg($extr), $_, $tarCode);
if ($tarCode !== 0) {
    fail("tar -xzf failed code=$tarCode");
}
$top = array_diff(scandir($extr), ['.', '..']);
sort($top);
$expected = ['bin', 'lib', 'nexo-plugin.toml'];
sort($expected);
if (array_values($top) !== $expected) {
    fail('unexpected top-level entries: ' . implode(',', $top));
}
if (!is_file("$extr/bin/" . PLUGIN_ID)) {
    fail('launcher missing: bin/' . PLUGIN_ID);
}
if (!is_file("$extr/lib/plugin/main.php")) {
    fail('main.php missing in lib/plugin/');
}
if (!is_file("$extr/lib/vendor/autoload.php")) {
    fail('vendor/autoload.php missing in lib/');
}
if (!is_file("$extr/lib/vendor/nexo/plugin-sdk/src/PluginAdapter.php")) {
    fail('vendored SDK missing in lib/vendor/nexo/plugin-sdk/');
}

// 9. Launcher executable bit.
$mode = fileperms("$extr/bin/" . PLUGIN_ID) & 0o777;
if ($mode !== 0o755) {
    fail(sprintf('launcher mode should be 0o755, got %o', $mode));
}

fwrite(STDOUT, "ok 1 - pack_tarball_produces_canonical_layout\n");
exit(0);
